### PR TITLE
Fix the potential deadlock issue of Ext2

### PR DESCRIPTION
**Pull request by liqinggd**

### CODE REVIEW

File: kernel/aster-nix/src/fs/ext2/inode.rs [hunk 1]
```diff
@@ -161,77 +164,79 @@ impl Inode {
...
-        let mut self_inner = self_inner.upgrade();
-        self_inner.remove_entry(name, 0)?;
-        drop(self_inner);
+        let (mut self_inner, mut dir_inner) = write_lock_two_inodes(self, &dir_inode);
+        // When we got the lock, the dir may have been modified by another thread
+        if self_inner.hard_links() == 0 {
+            return_errno_with_message!(Errno::ENOENT, "dir removed");
+        }
+        let (offset, new_ino) = self_inner
+            .get_entry(name, 0)
```
Comment (lucassong-mh) [inline kernel/aster-nix/src/fs/ext2/inode.rs:222]: A tiny readability issue here: When I first look at `get_entry()`, the 2nd parameter `offset` can be ambiguous, what is the relationship between that target dir entry and the offset, and what's "offset = 0" mean? Maybe we need a better encapsulation or some comments here.

File: kernel/aster-nix/src/fs/ext2/inode.rs [hunk 2]
```diff
@@ -161,77 +164,79 @@ impl Inode {
...
+            .map(|(offset, entry)| (offset, entry.ino()))
+            .ok_or(Error::new(Errno::ENOENT))?;
+        if dir_inode.ino != new_ino {
+            return_errno!(Errno::ENOENT);
+        }
+        let new_dir = self.fs().lookup_inode(dir_inode.ino)?;
```
Comment (lucassong-mh) [inline kernel/aster-nix/src/fs/ext2/inode.rs:228]: This `new_dir` could not be newly created right? If so, `potential_new_dir` is clearer.

File: kernel/aster-nix/src/fs/ext2/inode.rs [hunk 3]
```diff
@@ -583,6 +660,48 @@ impl Debug for Inode {
...
+    let mut guards: Vec<(usize, RwMutexWriteGuard<'a, Inner>)> = inodes
+        .into_iter()
+        .map(|(idx, inode)| (idx, inode.inner.write()))
+        .collect();
+
+    // Resort the guards by the input index.
+    // This ensures that the output order is consistent with the input.
+    guards.sort_by(|a, b| a.0.cmp(&b.0));
+    guards.into_iter().map(|(_, guard)| guard).collect()
+}
```
Comment (lucassong-mh) [inline kernel/aster-nix/src/fs/ext2/inode.rs:703]: A slightly better version:
```rust
fn write_lock_multiple_inodes<'a>(inodes: Vec<&'a Inode>) -> Vec<RwMutexWriteGuard<'a, Inner>> {
    let mut guards = vec![None; inodes.len()];

    let ordered_inodes: Vec<(usize, &Inode)> = inodes.into_iter().enumerate().collect();
    ordered_inodes.sort_by(|a, b| a.1.ino.cmp(&b.1.ino));
    
    for (original_idx, inode) in ordered_inodes {
        guards[original_idx] = Some(inode.inner.write());
    }
    guards.into_iter().map(|guard| guard.unwrap()).collect()
}
```

File: kernel/aster-nix/src/fs/ext2/inode.rs (hunk 1 above)
Comment (liqinggd) [reply kernel/aster-nix/src/fs/ext2/inode.rs]: I removed the argument, thanks for your comments.

File: kernel/aster-nix/src/fs/ext2/inode.rs (hunk 2 above)
Comment (liqinggd) [reply kernel/aster-nix/src/fs/ext2/inode.rs]: Thanks for your comments.

File: kernel/aster-nix/src/fs/ext2/inode.rs (hunk 3 above)
Comment (liqinggd) [reply kernel/aster-nix/src/fs/ext2/inode.rs:704]: Good suggestion!

**Comment by lucassong-mh** [comment]
LGTM

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!

**Pull request merged**